docs/environment_variables.rst
..
   This is a generated file; DO NOT EDIT!

   Please edit 'config/menu.json' to modify the list of variables and run
   './contrib/refresh-docs.sh' to regenerate this file

Environment variables
^^^^^^^^^^^^^^^^^^^^^

SATNOGS_API_TOKEN
~~~~~~~~~~~~~~~~~

:Type: *string*
:Default: *None*
:Required: *Yes*
:Description:
   SatNOGS Network API token associated with an account in SatNOGS Network.
   This token is secret.
   It can be found in SatNOGS Network user page.


SATNOGS_PRE_OBSERVATION_SCRIPT
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Type: *path*
:Default: *None*
:Required: *No*
:Description:
   A path to an executable to be executed before an observation job is started.
   Execution of this script blocks the observation job.


SATNOGS_POST_OBSERVATION_SCRIPT
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Type: *path*
:Default: *None*
:Required: *No*
:Description:
   A path to an executable to be executed after an observation job has finished.
   Execution of this script blocks the completion of an observation job.


SATNOGS_STATION_ID
~~~~~~~~~~~~~~~~~~

:Type: *integer*
:Default: *None*
:Required: *Yes*
:Description:
   The ID of the SatNOGS Network station this client will act as.
   The station must be owned by the user with the defined API token.


SATNOGS_STATION_LAT
~~~~~~~~~~~~~~~~~~~

:Type: *float*
:Default: *None*
:Required: *Yes*
:Description:
   Latitude of the station location.
   Higher precision of this value increases accuracy of Doppler correction while lower precision increases station location privacy.


SATNOGS_STATION_LON
~~~~~~~~~~~~~~~~~~~

:Type: *float*
:Default: *None*
:Required: *Yes*
:Description:
   Longitude of the station location.
   Higher precision of this value increases accuracy of Doppler correction while lower precision increases station location privacy.


SATNOGS_STATION_ELEV
~~~~~~~~~~~~~~~~~~~~

:Type: *integer*
:Default: *None*
:Required: *Yes*
:Description:
   Elevation of the station location.
   Higher precision of this value increases accuracy of Doppler correction while lower precision increases station location privacy.


SATNOGS_GPSD_CLIENT_ENABLED
~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Type: *boolean*
:Default: *False*
:Required: *No*
:Description:
   Enable SatNOGS Client to connect to a GPSd daemon to pull positional information.
   The position is queried once, during SatNOGS Client startup.


SATNOGS_GPSD_HOST
~~~~~~~~~~~~~~~~~

:Type: *host*
:Default: ``127.0.0.1``
:Required: *No*
:Description:
   Hostname or IP address of GPSd to connect to for pulling positional information.


SATNOGS_GPSD_PORT
~~~~~~~~~~~~~~~~~

:Type: *port*
:Default: ``2947``
:Required: *No*
:Description:
   Port of GPSd to connect to for pulling positional information.


SATNOGS_GPSD_TIMEOUT
~~~~~~~~~~~~~~~~~~~~

:Type: *integer*
:Default: ``0``
:Required: *No*
:Description:
   Time to wait until GPSd returns positional information.
   A value of 0 means to wait indefinitely.


SATNOGS_APP_PATH
~~~~~~~~~~~~~~~~

:Type: *path*
:Default: ``/tmp/.satnogs``
:Required: *No*
:Description:
   Base path for storing output files.


SATNOGS_OUTPUT_PATH
~~~~~~~~~~~~~~~~~~~

:Type: *path*
:Default: ``/tmp/.satnogs/data``
:Required: *No*
:Description:
   Path for storing "output files". Output files are:

     * Audio recordings (``satnogs_{obs_id}_{timestamp}.ogg``)
     * Waterfall Images (``waterfall_{obs_id}_{timestamp}.png``)
     * Decoded frames (``data_{obs_id}``), either raw, json-encoded

       or png image data from the APT demodulator

   Temporary (potentially incomplete) files are:

     * incomplete audio recordings (``receiving_satnogs_{obs_id}_{timestamp}.ogg``)
     * incomplete APT images (``receiving_data_{]obs_id}_{timestamp}.png``)
     * Waterfall data files (``receiving_waterfall_{]obs_id}_{timestamp}.dat``)

   SatNOGS Artifacts are handled seperately.


SATNOGS_COMPLETE_OUTPUT_PATH
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Type: *path*
:Default:
:Required: *No*
:Description:
   Path to move output files once they are completed.
   If set to empty, output files are deleted (default).
   Make sure to have enough storage available as the selected folder will
   continuously grow (about 30MB for a 5 minute observation).


SATNOGS_INCOMPLETE_OUTPUT_PATH
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Type: *path*
:Default: ``/tmp/.satnogs/data/incomplete``
:Required: *No*
:Description:
   Path for moving incomplete output files.


SATNOGS_REMOVE_RAW_FILES
~~~~~~~~~~~~~~~~~~~~~~~~

:Type: *boolean*
:Default: *True*
:Required: *No*
:Description:
   Remove raw data files used for generating waterfalls.


SATNOGS_KEEP_ARTIFACTS
~~~~~~~~~~~~~~~~~~~~~~

:Type: *boolean*
:Default: *False*
:Required: *No*
:Description:
   Do not remove artifact files after successfully uploading them.


SATNOGS_ARTIFACTS_OUTPUT_PATH
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Type: *path*
:Default: ``/tmp/.satnogs/artifacts``
:Required: *No*
:Description:
   Path for storing artifact files.


SATNOGS_VERIFY_SSL
~~~~~~~~~~~~~~~~~~

:Type: *boolean*
:Default: *True*
:Required: *No*
:Description:
   Verify SSL certificates for HTTPS requests.


SATNOGS_NETWORK_API_URL
~~~~~~~~~~~~~~~~~~~~~~~

:Type: *url*
:Default: ``https://network.satnogs.org/api/``
:Required: *No*
:Description:
   URL pointing to API of SatNOGS Network.


SATNOGS_NETWORK_API_QUERY_INTERVAL
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Type: *integer*
:Default: ``60``
:Required: *No*
:Description:
   Interval (in seconds) for pulling jobs form SatNOGS Network API.


SATNOGS_NETWORK_API_POST_INTERVAL
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Type: *integer*
:Default: ``180``
:Required: *No*
:Description:
   Interval (in seconds) for posting observation data to SatNOGS Network API.


SATNOGS_ROT_ENABLED
~~~~~~~~~~~~~~~~~~~

:Type: *boolean*
:Default: *True*
:Required: *No*
:Description:
   Enable SatNOGS Client to connect to an antenna rotator using the Hamlib library.


SATNOGS_ROT_MODEL
~~~~~~~~~~~~~~~~~

:Type: *string*
:Default: ``ROT_MODEL_DUMMY``
:Required: *No*
:Description:
   Rotator model to control.
   This value must be the model string of a Hamlib rotator.


SATNOGS_ROT_BAUD
~~~~~~~~~~~~~~~~

:Type: *integer*
:Default: ``19200``
:Required: *No*
:Description:
   Hamlib rotator serial interface baud rate.


SATNOGS_ROT_PORT
~~~~~~~~~~~~~~~~

:Type: *path*
:Default: ``/dev/ttyUSB0``
:Required: *No*
:Description:
   Path to Hamlib rotator serial port device.
   The device must be accessible to the user which SatNOGS Client is running.


SATNOGS_ROT_THRESHOLD
~~~~~~~~~~~~~~~~~~~~~

:Type: *integer*
:Default: ``4``
:Required: *No*
:Description:
   Azimuth/elevation threshold for moving the rotator.
   Position changes below this threshold will not cause the rotator to move.


SATNOGS_ROT_FLIP
~~~~~~~~~~~~~~~~

:Type: *boolean*
:Default: *False*
:Required: *No*
:Description:
   Enable rotator flipping during high elevation passes.


SATNOGS_ROT_FLIP_ANGLE
~~~~~~~~~~~~~~~~~~~~~~

:Type: *integer*
:Default: ``75``
:Required: *No*
:Description:
   Elevation angle above which the rotator will flip.


SATNOGS_RIG_IP
~~~~~~~~~~~~~~

:Type: *host*
:Default: ``127.0.0.1``
:Required: *No*
:Description:
   Hostname or IP address of Hamlib rotctld.


SATNOGS_RIG_PORT
~~~~~~~~~~~~~~~~

:Type: *integer*
:Default: ``4532``
:Required: *No*
:Description:
   Hamlib rigctld TCP port.


SATNOGS_SOAPY_RX_DEVICE
~~~~~~~~~~~~~~~~~~~~~~~

:Type: *string*
:Default: *None*
:Required: *Yes*
:Description:
   SoapySDR device driver to use for RX.
   This setting must be defined in the form ``driver=<name>`` where ``<name>`` is the name of the SoapySDR device driver to use.

   It also allows setting different devices for specific frequency ranges.

   The format for multiple devices is a space delimited list of frequency range -- device pairs:

   ``<min. freq.>-<max. freq>:driver=<name> <min. freq.>-<max. freq.>:driver=<name> ...``

   Where frequencies are in MHz and ``<name>`` is the name of the SoapySDR device driver to use for the specified frequency
   range.

   Example: two RTL-SDR receivers: one for VHF (135..148 MHz), another for UHF (430..470 MHz).

   ``135-148:driver=rtlsdr,serial=1 430-470:driver=rtlsdr,serial=2``


SATNOGS_RX_SAMP_RATE
~~~~~~~~~~~~~~~~~~~~

:Type: *integer*
:Default: *None*
:Required: *Yes*
:Description:
   SoapySDR device sample rate.
   Valid sample rates for attached devices can be queried using ``SoapySDRUtil --probe``.


SATNOGS_RX_BANDWIDTH
~~~~~~~~~~~~~~~~~~~~

:Type: *integer*
:Default: *Flowgraph-defined*
:Required: *No*
:Description:
   SoapySDR device RF bandwidth.
   This setting configures the RF filter on devices that support it.


SATNOGS_DOPPLER_CORR_PER_SEC
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Type: *integer*
:Default: *Flowgraph-defined*
:Required: *No*
:Description:
   Number of Doppler corrections per second requested by SatNOGS Radio.


SATNOGS_LO_OFFSET
~~~~~~~~~~~~~~~~~

:Type: *integer*
:Default: *Flowgraph-defined*
:Required: *No*
:Description:
   SoapySDR device local oscillator offset to apply.
   This setting is used to shift the carrier away from the DC spike.


SATNOGS_PPM_ERROR
~~~~~~~~~~~~~~~~~

:Type: *float*
:Default: *Flowgraph-defined*
:Required: *No*
:Description:
   SoapySDR device oscillator frequency error correction to apply.
   This setting is defined in parts per million.


SATNOGS_GAIN_MODE
~~~~~~~~~~~~~~~~~

:Type: *string*
:Default: ``Overall``
:Required: *No*
:Description:
   SoapySDR device gain mode.
   Valid values are: ``Overall``, ``Specific``, ``Settings Field``.


SATNOGS_RF_GAIN
~~~~~~~~~~~~~~~

:Type: *float*
:Default: *Flowgraph-defined*
:Required: *No*
:Description:
   SoapySDR device overall gain, in dB.
   Device drivers set individual, device specific gains to approximate linearity on the overall gain.


SATNOGS_ANTENNA
~~~~~~~~~~~~~~~

:Type: *string*
:Default: *None*
:Required: *Yes*
:Description:
   SoapySDR device antenna to use for RX.
   Valid antennas for attached devices can be queried using ``SoapySDRUtil --probe``.


SATNOGS_DEV_ARGS
~~~~~~~~~~~~~~~~

:Type: *string*
:Default: *Flowgraph-defined*
:Required: *No*
:Description:
   SoapySDR device arguments.
   Valid device arguments for attached devices can be queried using ``SoapySDRUtil --probe``.


SATNOGS_STREAM_ARGS
~~~~~~~~~~~~~~~~~~~

:Type: *string*
:Default: *Flowgraph-defined*
:Required: *No*
:Description:
   SoapySDR stream arguments.
   Valid stream arguments for attached devices can be queried using ``SoapySDRUtil --probe``.


SATNOGS_TUNE_ARGS
~~~~~~~~~~~~~~~~~

:Type: *string*
:Default: *Flowgraph-defined*
:Required: *No*
:Description:
   SoapySDR channel tune arguments.


SATNOGS_OTHER_SETTINGS
~~~~~~~~~~~~~~~~~~~~~~

:Type: *string*
:Default: *Flowgraph-defined*
:Required: *No*
:Description:
   SoapySDR channel other settings.


SATNOGS_DC_REMOVAL
~~~~~~~~~~~~~~~~~~

:Type: *boolean*
:Default: *Flowgraph-defined*
:Required: *No*
:Description:
   SoapySDR device automatic DC offset suppression.


SATNOGS_BB_FREQ
~~~~~~~~~~~~~~~

:Type: *string*
:Default: *Flowgraph-defined*
:Required: *No*
:Description:
   SoapySDR device baseband CORDIC frequency for devices that support it.


ENABLE_IQ_DUMP
~~~~~~~~~~~~~~

:Type: *boolean*
:Default: *False*
:Required: *No*
:Description:
   Create I/Q data dumps for every observation.
   Use this feature with caution.
   Enabling this setting will store large amount of data on the filesystem.


IQ_DUMP_FILENAME
~~~~~~~~~~~~~~~~

:Type: *path*
:Default: *None*
:Required: *No*
:Description:
   Path to file for storing I/Q data dumps.


DISABLE_DECODED_DATA
~~~~~~~~~~~~~~~~~~~~

:Type: *boolean*
:Default: *False*
:Required: *No*
:Description:
   Disable output of decoded data.


UDP_DUMP_HOST
~~~~~~~~~~~~~

:Type: *host*
:Default: *Flowgraph-defined*
:Required: *No*
:Description:
   IP destination of UDP data with Doppler corrected I/Q.


UDP_DUMP_PORT
~~~~~~~~~~~~~

:Type: *port*
:Default: ``57356``
:Required: *No*
:Description:
   Port for UDP data with Doppler corrected I/Q.


SATNOGS_UPLOAD_AUDIO_FILES
~~~~~~~~~~~~~~~~~~~~~~~~~~

:Type: *boolean*
:Default: *True*
:Required: *No*
:Description:
   Enable/Disable uploading audio files to SatNOGS network.


SATNOGS_UPLOAD_WATERFALL_FILES
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Type: *boolean*
:Default: *True*
:Required: *No*
:Description:
   Enable/Disable uploading waterfalls to SatNOGS network.


SATNOGS_WATERFALL_AUTORANGE
~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Type: *boolean*
:Default: *True*
:Required: *No*
:Description:
   Automatically set power level range of waterfall images.


SATNOGS_WATERFALL_MIN_VALUE
~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Type: *integer*
:Default: ``-100``
:Required: *No*
:Description:
   Minimum power level of waterfall images.


SATNOGS_WATERFALL_MAX_VALUE
~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Type: *integer*
:Default: ``-50``
:Required: *No*
:Description:
   Maximum power level of waterfall images.


SATNOGS_ARTIFACTS_ENABLED
~~~~~~~~~~~~~~~~~~~~~~~~~

:Type: *boolean*
:Default: *False*
:Required: *No*
:Description:
   Enable generation and uploading of HDF5 artifacts files to SatNOGS DB.


SATNOGS_ARTIFACTS_API_URL
~~~~~~~~~~~~~~~~~~~~~~~~~

:Type: *url*
:Default: ``https://db.satnogs.org/api/``
:Required: *No*
:Description:
   URL pointing to API of SatNOGS DB for uploading artifacts.


SATNOGS_ARTIFACTS_API_POST_INTERVAL
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Type: *integer*
:Default: ``180``
:Required: *No*
:Description:
   Interval (in seconds) for posting artifacts to SatNOGS DB.


SATNOGS_ARTIFACTS_API_TOKEN
~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Type: *string*
:Default: *None*
:Required: *No*
:Description:
   SatNOGS DB API token associated with an account in SatNOGS DB.
   This token is secret.
   It is used to upload artifacts to SatNOGS DB.
   It can be found in SatNOGS DB user page.


SATNOGS_LOG_LEVEL
~~~~~~~~~~~~~~~~~

:Type: *string*
:Default: ``WARNING``
:Required: *No*
:Description:
   SatNOGS Client logging level.
   Valid values are:

     * ``CRITICAL``
     * ``ERROR``
     * ``WARNING``
     * ``INFO``
     * ``DEBUG``


SATNOGS_SCHEDULER_LOG_LEVEL
~~~~~~~~~~~~~~~~~~~~~~~~~~~

:Type: *string*
:Default: ``WARNING``
:Required: *No*
:Description:
   SatNOGS Client scheduler logging level.
   Valid values are:

     * ``CRITICAL``
     * ``ERROR``
     * ``WARNING``
     * ``INFO``
     * ``DEBUG``


SENTRY_DSN
~~~~~~~~~~

:Type: *string*
:Default: ``d50342fb75aa8f3945e2f846b77a0cdb7c7d2275``
:Required: *No*
:Description:
   Sentry Data Source Name used for sending events to application monitoring and error tracking server.


SENTRY_ENABLED
~~~~~~~~~~~~~~

:Type: *boolean*
:Default: *False*
:Required: *No*
:Description:
   Enable sending events to Sentry application monitoring and error tracking server.


satnogsclient/settings.py
"""SatNOGS Client settings file"""
from __future__ import absolute_import, division, print_function

import os
from distutils.util import strtobool  # pylint: disable=E0401,E0611
from os import environ

from dotenv import load_dotenv
from validators.url import url


def _cast_or_none(func, value):
    try:
        return func(value)
    except (ValueError, TypeError):
        return None


try:
    load_dotenv()
except EnvironmentError:
    # XXX: Workaround for readthedocs  # pylint: disable=fixme
    pass

# Ground station information
SATNOGS_API_TOKEN = environ.get('SATNOGS_API_TOKEN', None)
SATNOGS_PRE_OBSERVATION_SCRIPT = environ.get('SATNOGS_PRE_OBSERVATION_SCRIPT', None)
SATNOGS_POST_OBSERVATION_SCRIPT = environ.get('SATNOGS_POST_OBSERVATION_SCRIPT', None)
SATNOGS_STATION_ID = _cast_or_none(int, environ.get('SATNOGS_STATION_ID', None))
SATNOGS_STATION_LAT = _cast_or_none(float, environ.get('SATNOGS_STATION_LAT', None))
SATNOGS_STATION_LON = _cast_or_none(float, environ.get('SATNOGS_STATION_LON', None))
SATNOGS_STATION_ELEV = _cast_or_none(int, environ.get('SATNOGS_STATION_ELEV', None))
SATNOGS_GPSD_CLIENT_ENABLED = bool(strtobool(environ.get('SATNOGS_GPSD_CLIENT_ENABLED', 'False')))
SATNOGS_GPSD_HOST = environ.get('SATNOGS_GPSD_HOST', '127.0.0.1')
SATNOGS_GPSD_PORT = _cast_or_none(int, environ.get('SATNOGS_GPSD_PORT', 2947))
SATNOGS_GPSD_TIMEOUT = _cast_or_none(int, environ.get('SATNOGS_GPSD_TIMEOUT', 30))

# Output paths
SATNOGS_APP_PATH = environ.get('SATNOGS_APP_PATH', '/tmp/.satnogs')
SATNOGS_OUTPUT_PATH = environ.get('SATNOGS_OUTPUT_PATH', '/tmp/.satnogs/data')
SATNOGS_COMPLETE_OUTPUT_PATH = environ.get('SATNOGS_COMPLETE_OUTPUT_PATH', '')
SATNOGS_INCOMPLETE_OUTPUT_PATH = environ.get('SATNOGS_INCOMPLETE_OUTPUT_PATH',
                                             '/tmp/.satnogs/data/incomplete')
SATNOGS_ARTIFACTS_OUTPUT_PATH = environ.get('SATNOGS_ARTIFACTS_OUTPUT_PATH',
                                            '/tmp/.satnogs/artifacts')

for p in (SATNOGS_APP_PATH, SATNOGS_OUTPUT_PATH, SATNOGS_COMPLETE_OUTPUT_PATH,
          SATNOGS_INCOMPLETE_OUTPUT_PATH):
    if p and not os.path.exists(p):
        os.mkdir(p)

SATNOGS_REMOVE_RAW_FILES = bool(strtobool(environ.get('SATNOGS_REMOVE_RAW_FILES', 'True')))
SATNOGS_KEEP_ARTIFACTS = bool(strtobool(environ.get('SATNOGS_KEEP_ARTIFACTS', 'False')))

SATNOGS_VERIFY_SSL = bool(strtobool(environ.get('SATNOGS_VERIFY_SSL', 'True')))

SATNOGS_NETWORK_API_URL = environ.get('SATNOGS_NETWORK_API_URL',
                                      'https://network.satnogs.org/api/')
SATNOGS_NETWORK_API_QUERY_INTERVAL = _cast_or_none(int,
                                                   environ.get(
                                                       'SATNOGS_NETWORK_API_QUERY_INTERVAL',
                                                       60))  # In seconds
SATNOGS_NETWORK_API_POST_INTERVAL = _cast_or_none(int,
                                                  environ.get('SATNOGS_NETWORK_API_POST_INTERVAL',
                                                              180))  # In seconds
SATNOGS_NETWORK_API_TIMEOUT = 1800  # In seconds
GNURADIO_UDP_PORT = 16886
GNURADIO_IP = '127.0.0.1'

SATNOGS_ROT_ENABLED = bool(strtobool(environ.get('SATNOGS_ROT_ENABLED', 'True')))
SATNOGS_ROT_MODEL = environ.get('SATNOGS_ROT_MODEL', 'ROT_MODEL_DUMMY')
SATNOGS_ROT_BAUD = int(environ.get('SATNOGS_ROT_BAUD', 19200))
SATNOGS_ROT_PORT = environ.get('SATNOGS_ROT_PORT', '/dev/ttyUSB0')
SATNOGS_RIG_IP = environ.get('SATNOGS_RIG_IP', '127.0.0.1')
SATNOGS_RIG_PORT = int(environ.get('SATNOGS_RIG_PORT', 4532))
SATNOGS_ROT_THRESHOLD = int(environ.get('SATNOGS_ROT_THRESHOLD', 4))
SATNOGS_ROT_FLIP = bool(strtobool(environ.get('SATNOGS_ROT_FLIP', 'False')))
SATNOGS_ROT_FLIP_ANGLE = int(environ.get('SATNOGS_ROT_FLIP_ANGLE', 75))

# Common script parameters
SATNOGS_SOAPY_RX_DEVICE = environ.get('SATNOGS_SOAPY_RX_DEVICE', None)
SATNOGS_RX_SAMP_RATE = environ.get('SATNOGS_RX_SAMP_RATE', None)
SATNOGS_RX_BANDWIDTH = environ.get('SATNOGS_RX_BANDWIDTH', None)
SATNOGS_DOPPLER_CORR_PER_SEC = environ.get('SATNOGS_DOPPLER_CORR_PER_SEC', None)
SATNOGS_LO_OFFSET = environ.get('SATNOGS_LO_OFFSET', None)
SATNOGS_PPM_ERROR = environ.get('SATNOGS_PPM_ERROR', None)
SATNOGS_GAIN_MODE = environ.get('SATNOGS_GAIN_MODE', 'Overall')
SATNOGS_RF_GAIN = environ.get('SATNOGS_RF_GAIN', None)
SATNOGS_ANTENNA = environ.get('SATNOGS_ANTENNA', None)
SATNOGS_DEV_ARGS = environ.get('SATNOGS_DEV_ARGS', None)
SATNOGS_STREAM_ARGS = environ.get('SATNOGS_STREAM_ARGS', None)
SATNOGS_TUNE_ARGS = environ.get('SATNOGS_TUNE_ARGS', None)
SATNOGS_OTHER_SETTINGS = environ.get('SATNOGS_OTHER_SETTINGS', None)
SATNOGS_DC_REMOVAL = environ.get('SATNOGS_DC_REMOVAL', None)
SATNOGS_BB_FREQ = environ.get('SATNOGS_BB_FREQ', None)

ENABLE_IQ_DUMP = bool(strtobool(environ.get('ENABLE_IQ_DUMP', 'False')))
IQ_DUMP_FILENAME = environ.get('IQ_DUMP_FILENAME', None)
DISABLE_DECODED_DATA = bool(strtobool(environ.get('DISABLE_DECODED_DATA', 'False')))
UDP_DUMP_HOST = environ.get('UDP_DUMP_HOST', None)
UDP_DUMP_PORT = int(environ.get('UDP_DUMP_PORT', 57356))

# Upload settings
SATNOGS_UPLOAD_AUDIO_FILES = bool(strtobool(environ.get('SATNOGS_UPLOAD_AUDIO_FILES', 'True')))
SATNOGS_UPLOAD_WATERFALL_FILES = bool(
    strtobool(environ.get('SATNOGS_UPLOAD_WATERFALL_FILES', 'True')))

# Artifacts settings
ARTIFACTS_ENABLED = bool(strtobool(environ.get('SATNOGS_ARTIFACTS_ENABLED', 'False')))
ARTIFACTS_API_URL = environ.get('SATNOGS_ARTIFACTS_API_URL', 'https://db.satnogs.org/api/')
ARTIFACTS_API_POST_INTERVAL = _cast_or_none(int,
                                            environ.get('SATNOGS_ARTIFACTS_API_POST_INTERVAL',
                                                        180))  # In seconds
ARTIFACTS_API_TIMEOUT = 1800  # In seconds
ARTIFACTS_API_TOKEN = environ.get('SATNOGS_ARTIFACTS_API_TOKEN', None)

# Waterfall settings
SATNOGS_WATERFALL_AUTORANGE = bool(strtobool(environ.get('SATNOGS_WATERFALL_AUTORANGE', 'True')))
SATNOGS_WATERFALL_MIN_VALUE = int(environ.get('SATNOGS_WATERFALL_MIN_VALUE', -100))
SATNOGS_WATERFALL_MAX_VALUE = int(environ.get('SATNOGS_WATERFALL_MAX_VALUE', -50))

# Logging configuration
LOG_FORMAT = '{levelname:<8s} {name:<31s} {message}'
LOG_LEVEL = environ.get('SATNOGS_LOG_LEVEL', 'WARNING')
SCHEDULER_LOG_LEVEL = environ.get('SATNOGS_SCHEDULER_LOG_LEVEL', 'WARNING')

# Sentry
SENTRY_DSN = environ.get('SENTRY_DSN',
                         'https://5e18316a65b1457c99fe81bf869e26d6@sentry.io/2818206')
SENTRY_ENABLED = bool(strtobool(environ.get('SENTRY_ENABLED', 'False')))

REQUIRED_VARIABLES = [
    'SATNOGS_API_TOKEN', 'SATNOGS_STATION_ID', 'SATNOGS_SOAPY_RX_DEVICE', 'SATNOGS_RX_SAMP_RATE',
    'SATNOGS_ANTENNA'
]


def validate(logger):
    """Validate the provided settings:
    - Check for the existance of all required variables
    - Validate format of the provided value for some required variables

    Since this module has to be loaded before the logger has been initialized,
    this method requires a configured logger to be passed.

    Arguments:
    logger -- the output logger
    """
    settings_valid = True

    # Get all variable in global scobe (this includes all global variables from this module)
    settings = globals()

    for variable_name in REQUIRED_VARIABLES:
        # Check the value of the variable defined in settings
        if not settings[variable_name]:
            logger.error('%s not configured but required', variable_name)
            settings_valid = False

    try:
        url(SATNOGS_NETWORK_API_URL)
    except ValueError:
        logger.error('Invalid SATNOGS_NETWORK_API_URL: %s', SATNOGS_NETWORK_API_URL)
        settings_valid = False

    if not (SATNOGS_STATION_LAT or SATNOGS_GPSD_CLIENT_ENABLED):
        logger.error('SATNOGS_STATION_LAT not configured')
        settings_valid = False

    if not (SATNOGS_STATION_LON or SATNOGS_GPSD_CLIENT_ENABLED):
        logger.error('SATNOGS_STATION_LON not configured')
        settings_valid = False

    if not (SATNOGS_STATION_ELEV or SATNOGS_GPSD_CLIENT_ENABLED):
        logger.error('SATNOGS_STATION_ELEV not configured')
        settings_valid = False

    return settings_valid
